src/openrct2/windows/textreader.c

```diff
@@ -0,0 +1,281 @@
+ *
+ * OpenRCT2 is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * A full copy of the GNU General Public License can be found in licence.txt
+ *****************************************************************************/
+#pragma endregion
+
+#include "../interface/themes.h"
+#include "../interface/viewport.h"
+#include "../interface/widget.h"
+#include "../interface/window.h"
+#include "../localisation/localisation.h"
+#include "../OpenRCT2.h"
+#include "../platform/platform.h"
+#include "../rct2.h"
+#include "../util/util.h"
+#include "../world/footpath.h"
+#include "../world/map.h"
+#include "../world/scenery.h"
+
+enum {
+	WIDX_BACKGROUND,
+	WIDX_TITLE,
+	WIDX_CLOSE,
+	WIDX_CONTENT_PANEL,
+	WIDX_SCROLL
+};
+
+#define WW 500
+#define WH 400
+#define MIN_WW 300
+#define MIN_WH 200
+
+rct_widget window_textreader_widgets[] = {
+	{ WWT_FRAME,			0,	0,			WW - 1,	0,		WH - 1,		0xFFFFFFFF,						STR_NONE },				// panel / background
+	{ WWT_CAPTION,			0,	1,			WW - 2,	1,		14,			STR_TEXTREADER_TITLE,			STR_WINDOW_TITLE_TIP },	// title bar
+	{ WWT_CLOSEBOX,			0,	WW - 13,	WW - 3,	2,		13,			STR_CLOSE_X,					STR_CLOSE_WINDOW_TIP },	// close x button
+	{ WWT_RESIZE,			1,	0,			WW - 1,	14,		WH - 1,		0xFFFFFFFF,						STR_NONE },				// content panel
+	{ WWT_SCROLL,			1,	3,			WW - 3,	16,		WH - 15,	SCROLL_BOTH,								STR_NONE },				// scroll area
+	{ WIDGETS_END },
+};
+
+static void window_textreader_close(rct_window *w);
+static void window_textreader_mouseup(rct_window *w, sint32 widgetIndex);
+static void window_textreader_resize(rct_window *w);
+static void window_textreader_scrollgetsize(rct_window *w, sint32 scrollIndex, sint32 *width, sint32 *height);
+static void window_textreader_invalidate(rct_window *w);
+static void window_textreader_paint(rct_window *w, rct_drawpixelinfo *dpi);
+static void window_textreader_scrollpaint(rct_window *w, rct_drawpixelinfo *dpi, sint32 scrollIndex);
+
+static rct_window_event_list window_textreader_events = {
+	window_textreader_close,
+	window_textreader_mouseup,
+	window_textreader_resize,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	window_textreader_scrollgetsize,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	window_textreader_invalidate,
+	window_textreader_paint,
+	window_textreader_scrollpaint
+};
+
+static bool window_textreader_read_file();
+static void window_textreader_dispose_file();
+
+static char *_textreaderText = NULL;
+static size_t _textreaderTextSize = 0;
+static char **_textreaderLines = NULL;
+static sint32 _textreaderNumLines = 0;
+static sint32 _textreaderLongestLineWidth = 0;
+char _textreaderFilePath[MAX_PATH];
+
+rct_window *window_textreader_open(const char * file_path)
```

Comment:
There should be no file operations in `textreader` window. It should only receive the buffer and its size. If it needs figuring out longest line size, it should do so itself.